File: ios/AgenticInbox/Inboxies/Services/SearchQueryParser.swift
import Foundation

/// Parsed Gmail-style search operators — mirrors `app/lib/search-parser.ts`.
struct ParsedSearch: Equatable {
    var query: String = ""
    var from: String?
    var to: String?
    var subject: String?
    var folder: String?
    var isRead: Bool?
    var isStarred: Bool?
    var hasAttachment: Bool?
    var dateStart: String?
    var dateEnd: String?

    /// True when any structured filter is set (free-text may still be empty).
    var hasStructuredFilters: Bool {
        from != nil
            || to != nil
            || subject != nil
            || folder != nil
            || isRead != nil
            || isStarred != nil
            || hasAttachment == true
            || dateStart != nil
            || dateEnd != nil
    }

    /// Query params for `GET …/search`, omitting empty values.
    var apiQueryItems: [String: String] {
        var params: [String: String] = [:]
        if !query.isEmpty { params["query"] = query }
        if let from { params["from"] = from }
        if let to { params["to"] = to }
        if let subject { params["subject"] = subject }
        if let folder { params["folder"] = folder }
        if let dateStart { params["date_start"] = dateStart }
        if let dateEnd { params["date_end"] = dateEnd }
        if let isRead { params["is_read"] = isRead ? "true" : "false" }
        if let isStarred { params["is_starred"] = isStarred ? "true" : "false" }
        if hasAttachment == true { params["has_attachment"] = "true" }
        return params
    }
}

enum SearchQueryParser {
    /// Matches `operator:value` or `operator:"quoted value"` (case-insensitive ops).
    private static let operatorRegex: NSRegularExpression = {
        try! NSRegularExpression(
            pattern: #"\b(from|to|subject|in|is|has|before|after):(?:"([^"]*?)"|(\S+))"#,
            options: [.caseInsensitive]
        )
    }()

    static func parse(_ input: String) -> ParsedSearch {
        var result = ParsedSearch()
        let nsRange = NSRange(input.startIndex..<input.endIndex, in: input)
        let matches = operatorRegex.matches(in: input, options: [], range: nsRange)

        var remaining = input
        var ops: [(op: String, value: String)] = []

        for match in matches {
            guard match.numberOfRanges >= 4,
                  let fullRange = Range(match.range(at: 0), in: input),
                  let opRange = Range(match.range(at: 1), in: input)
            else { continue }

            let value: String
            if match.range(at: 2).location != NSNotFound,
               let quotedRange = Range(match.range(at: 2), in: input) {
                value = String(input[quotedRange])
            } else if match.range(at: 3).location != NSNotFound,
                      let bareRange = Range(match.range(at: 3), in: input) {
                value = String(input[bareRange])
            } else {
                value = ""
            }

            let op = String(input[opRange]).lowercased()
            ops.append((op, value))

            let fullMatch = String(input[fullRange])
            if let found = remaining.range(of: fullMatch) {
                remaining.removeSubrange(found)
            }
        }

        result.query = remaining
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)

        for (op, value) in ops {
            switch op {
            case "from":
                result.from = value
            case "to":
                result.to = value
            case "subject":
                result.subject = value
            case "in":
                result.folder = value.lowercased()
            case "is":
                switch value.lowercased() {
                case "unread": result.isRead = false
                case "read": result.isRead = true
                case "starred": result.isStarred = true
                case "unstarred": result.isStarred = false
                default: break
                }
            case "has":
                if value.lowercased() == "attachment" {
                    result.hasAttachment = true
                }
            case "before":
                result.dateEnd = normalizeDate(value)
            case "after":
                result.dateStart = normalizeDate(value)
            default:
                break
            }
        }

        return result
    }

    /// Normalize to ISO-8601. Accepts YYYY-MM-DD or ISO date-times.
    private static func normalizeDate(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        let out = ISO8601DateFormatter()
        out.formatOptions = [.withInternetDateTime]

        let withFraction = ISO8601DateFormatter()
        withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = withFraction.date(from: trimmed) {
            return out.string(from: d)
        }
        if let d = out.date(from: trimmed) {
            return out.string(from: d)
        }

        let day = DateFormatter()
        day.calendar = Calendar(identifier: .gregorian)
        day.locale = Locale(identifier: "en_US_POSIX")
        day.timeZone = TimeZone(secondsFromGMT: 0)
        day.dateFormat = "yyyy-MM-dd"
        if let d = day.date(from: trimmed) {
            return out.string(from: d)
        }

        return nil
    }
}

File: android/app/src/main/java/co/inboxies/app/util/SearchQueryParser.kt
package co.inboxies.app.util

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.format.DateTimeParseException

/**
 * Parsed Gmail-style search operators — mirrors `app/lib/search-parser.ts`.
 */
data class ParsedSearch(
    val query: String = "",
    val from: String? = null,
    val to: String? = null,
    val subject: String? = null,
    val folder: String? = null,
    val isRead: Boolean? = null,
    val isStarred: Boolean? = null,
    val isReplyLater: Boolean? = null,
    val hasAttachment: Boolean? = null,
    val dateStart: String? = null,
    val dateEnd: String? = null,
) {
    val hasStructuredFilters: Boolean
        get() = from != null ||
            to != null ||
            subject != null ||
            folder != null ||
            isRead != null ||
            isStarred != null ||
            isReplyLater != null ||
            hasAttachment == true ||
            dateStart != null ||
            dateEnd != null

    /** Query params for `GET …/search`, omitting empty values. */
    fun toApiQuery(): Map<String, String> = buildMap {
        if (query.isNotEmpty()) put("query", query)
        from?.let { put("from", it) }
        to?.let { put("to", it) }
        subject?.let { put("subject", it) }
        folder?.let { put("folder", it) }
        dateStart?.let { put("date_start", it) }
        dateEnd?.let { put("date_end", it) }
        isRead?.let { put("is_read", if (it) "true" else "false") }
        isStarred?.let { put("is_starred", if (it) "true" else "false") }
        isReplyLater?.let { put("is_reply_later", if (it) "true" else "false") }
        if (hasAttachment == true) put("has_attachment", "true")
    }
}

object SearchQueryParser {
    private val operatorRegex =
        Regex("""\b(from|to|subject|in|is|has|before|after):(?:"([^"]*?)"|(\S+))""", RegexOption.IGNORE_CASE)

    fun parse(input: String): ParsedSearch {
        val matches = operatorRegex.findAll(input).toList()
        var remaining = input
        for (match in matches) {
            remaining = remaining.replaceFirst(match.value, "")
        }
        val freeText = remaining.replace(Regex("""\s+"""), " ").trim()

        var from: String? = null
        var to: String? = null
        var subject: String? = null
        var folder: String? = null
        var isRead: Boolean? = null
        var isStarred: Boolean? = null
        var isReplyLater: Boolean? = null
        var hasAttachment: Boolean? = null
        var dateStart: String? = null
        var dateEnd: String? = null

        for (match in matches) {
            val op = match.groupValues[1].lowercase()
            val value = match.groupValues[2].ifEmpty { match.groupValues[3] }
            when (op) {
                "from" -> from = value
                "to" -> to = value
                "subject" -> subject = value
                "in" -> folder = value.lowercase()
                "is" -> when (value.lowercase()) {
                    "unread" -> isRead = false
                    "read" -> isRead = true
                    "starred" -> isStarred = true
                    "unstarred" -> isStarred = false
                    "reply-later", "reply_later" -> isReplyLater = true
                }
                "has" -> if (value.lowercase() == "attachment") hasAttachment = true
                "before" -> dateEnd = normalizeDate(value)
                "after" -> dateStart = normalizeDate(value)
            }
        }

        return ParsedSearch(
            query = freeText,
            from = from,
            to = to,
            subject = subject,
            folder = folder,
            isRead = isRead,
            isStarred = isStarred,
            isReplyLater = isReplyLater,
            hasAttachment = hasAttachment,
            dateStart = dateStart,
            dateEnd = dateEnd,
        )
    }

    /**
     * Normalize to ISO-8601. Mirrors web `new Date(value).toISOString()` for
     * single-token values: ISO instants, `YYYY-MM-DD`, and `M/D/YYYY`.
     */
    private fun normalizeDate(value: String): String? {
        val trimmed = value.trim()
        if (trimmed.isEmpty()) return null
        return try {
            Instant.parse(trimmed).toString()
        } catch (_: DateTimeParseException) {
            try {
                LocalDate.parse(trimmed).atStartOfDay(ZoneOffset.UTC).toInstant().toString()
            } catch (_: DateTimeParseException) {
                parseSlashDate(trimmed)
            }
        }
    }

    private fun parseSlashDate(value: String): String? {
        val parts = value.split('/')
        if (parts.size != 3) return null
        val month = parts[0].toIntOrNull() ?: return null
        val day = parts[1].toIntOrNull() ?: return null
        val year = parts[2].toIntOrNull() ?: return null
        return try {
            LocalDate.of(year, month, day).atStartOfDay(ZoneOffset.UTC).toInstant().toString()
        } catch (_: Exception) {
            null
        }
    }
}
